File: step_04_binary_search/4.1_bs_on_1d_arrays/P005_floor_and_ceil_in_sorted_array/Solution.java
/**
 * Problem: Floor and Ceil in Sorted Array
 * Difficulty: EASY | XP: 10
 *
 * Find the floor (largest <= target) and ceil (smallest >= target)
 * in a sorted array using binary search.
 *
 * @author DSA_Nova
 */
import java.util.Arrays;

// ============================================================
// Approach 1: Brute Force (Linear Scan)
// Time: O(n) | Space: O(1)
// ============================================================
class BruteForce {
    /**
     * Scan left to right. Last element <= target is floor.
     * First element >= target is ceil.
     */
    public static int[] floorCeil(int[] nums, int target) {
        int floor = -1, ceil = -1;
        for (int num : nums) {
            if (num <= target) {
                floor = num;  // keep overwriting -- last one is largest <= target
            }
            if (num >= target && ceil == -1) {
                ceil = num;   // first one is smallest >= target
            }
        }
        return new int[]{floor, ceil};
    }
}

// ============================================================
// Approach 2: Optimal (Two Separate Binary Searches)
// Time: O(log n) | Space: O(1)
// ============================================================
class Optimal {
    /**
     * Run one binary search for floor and one for ceil.
     * Floor: record and go right. Ceil: record and go left.
     */
    public static int[] floorCeil(int[] nums, int target) {
        return new int[]{findFloor(nums, target), findCeil(nums, target)};
    }

    private static int findFloor(int[] nums, int target) {
        int lo = 0, hi = nums.length - 1;
        int floor = -1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] <= target) {
                floor = nums[mid];  // candidate -- try to find larger
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
        return floor;
    }

    private static int findCeil(int[] nums, int target) {
        int lo = 0, hi = nums.length - 1;
        int ceil = -1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] >= target) {
                ceil = nums[mid];  // candidate -- try to find smaller
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        }
        return ceil;
    }
}

// ============================================================
// Approach 3: Best (Single-Pass Binary Search)
// Time: O(log n) | Space: O(1)
// ============================================================
class Best {
    /**
     * One binary search pass that updates both floor and ceil candidates.
     * Elements < target are floor candidates; elements > target are ceil candidates.
     * If we hit target exactly, both floor and ceil = target.
     */
    public static int[] floorCeil(int[] nums, int target) {
        int lo = 0, hi = nums.length - 1;
        int floor = -1, ceil = -1;

        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] == target) {
                return new int[]{target, target};  // exact match
            } else if (nums[mid] < target) {
                floor = nums[mid];  // best floor candidate so far
                lo = mid + 1;
            } else {
                ceil = nums[mid];   // best ceil candidate so far
                hi = mid - 1;
            }
        }
        return new int[]{floor, ceil};
    }
}

// Main driver
public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Floor and Ceil in Sorted Array ===\n");

        int[][] arrays = {
            {1, 2, 8, 10, 10, 12, 19},
            {1, 2, 8, 10, 10, 12, 19},
            {1, 2, 8, 10, 10, 12, 19},
            {1, 2, 8, 10, 10, 12, 19},
            {5},
            {5},
            {3, 7},
        };
        int[] targets  = {5, 10, 0, 25, 5, 3, 5};
        int[][] expected = {
            {2, 8},
            {10, 10},
            {-1, 1},
            {19, -1},
            {5, 5},
            {-1, 5},
            {3, 7},
        };

        for (int t = 0; t < arrays.length; t++) {
            int[] b = BruteForce.floorCeil(arrays[t], targets[t]);
            int[] o = Optimal.floorCeil(arrays[t], targets[t]);
            int[] s = Best.floorCeil(arrays[t], targets[t]);
            boolean pass = Arrays.equals(b, expected[t])
                        && Arrays.equals(o, expected[t])
                        && Arrays.equals(s, expected[t]);

            System.out.println("Input:    " + Arrays.toString(arrays[t]) + ", target=" + targets[t]);
            System.out.println("Brute:    floor=" + b[0] + ", ceil=" + b[1]);
            System.out.println("Optimal:  floor=" + o[0] + ", ceil=" + o[1]);
            System.out.println("Best:     floor=" + s[0] + ", ceil=" + s[1]);
            System.out.println("Expected: floor=" + expected[t][0] + ", ceil=" + expected[t][1]);
            System.out.println("Pass:     " + pass);
            System.out.println();
        }
    }
}

File: step_04_binary_search/4.1_bs_on_1d_arrays/P005_floor_and_ceil_in_sorted_array/solution.py
"""
Problem: Floor and Ceil in Sorted Array
Difficulty: EASY | XP: 10

Find the floor (largest <= target) and ceil (smallest >= target)
in a sorted array using binary search.
"""
from typing import List, Tuple


# ============================================================
# APPROACH 1: BRUTE FORCE (Linear Scan)
# Time: O(n) | Space: O(1)
# ============================================================
def brute_force(nums: List[int], target: int) -> Tuple[int, int]:
    """Scan left to right. Last <= target is floor, first >= target is ceil."""
    floor, ceil = -1, -1
    for num in nums:
        if num <= target:
            floor = num   # keep overwriting -- last is largest <= target
        if num >= target and ceil == -1:
            ceil = num    # first is smallest >= target
    return (floor, ceil)


# ============================================================
# APPROACH 2: OPTIMAL (Two Separate Binary Searches)
# Time: O(log n) | Space: O(1)
# ============================================================
def optimal(nums: List[int], target: int) -> Tuple[int, int]:
    """Run one binary search for floor and one for ceil."""
    return (find_floor(nums, target), find_ceil(nums, target))


def find_floor(nums: List[int], target: int) -> int:
    """Find largest element <= target. Record candidate and go right."""
    lo, hi = 0, len(nums) - 1
    floor = -1
    while lo <= hi:
        mid = lo + (hi - lo) // 2
        if nums[mid] <= target:
            floor = nums[mid]  # candidate -- try to find larger
            lo = mid + 1
        else:
            hi = mid - 1
    return floor


def find_ceil(nums: List[int], target: int) -> int:
    """Find smallest element >= target. Record candidate and go left."""
    lo, hi = 0, len(nums) - 1
    ceil = -1
    while lo <= hi:
        mid = lo + (hi - lo) // 2
        if nums[mid] >= target:
            ceil = nums[mid]  # candidate -- try to find smaller
            hi = mid - 1
        else:
            lo = mid + 1
    return ceil


# ============================================================
# APPROACH 3: BEST (Single-Pass Binary Search)
# Time: O(log n) | Space: O(1)
# ============================================================
def best(nums: List[int], target: int) -> Tuple[int, int]:
    """One binary search updating both floor and ceil candidates."""
    lo, hi = 0, len(nums) - 1
    floor, ceil = -1, -1

    while lo <= hi:
        mid = lo + (hi - lo) // 2
        if nums[mid] == target:
            return (target, target)  # exact match
        elif nums[mid] < target:
            floor = nums[mid]  # best floor candidate so far
            lo = mid + 1
        else:
            ceil = nums[mid]   # best ceil candidate so far
            hi = mid - 1

    return (floor, ceil)


# ============================================================
# DRIVER
# ============================================================
if __name__ == "__main__":
    print("=== Floor and Ceil in Sorted Array ===\n")

    test_cases = [
        ([1, 2, 8, 10, 10, 12, 19], 5,  (2, 8)),
        ([1, 2, 8, 10, 10, 12, 19], 10, (10, 10)),
        ([1, 2, 8, 10, 10, 12, 19], 0,  (-1, 1)),
        ([1, 2, 8, 10, 10, 12, 19], 25, (19, -1)),
        ([5], 5, (5, 5)),
        ([5], 3, (-1, 5)),
        ([3, 7], 5, (3, 7)),
    ]

    for nums, target, expected in test_cases:
        b = brute_force(nums, target)
        o = optimal(nums, target)
        s = best(nums, target)
        status = "PASS" if b == expected and o == expected and s == expected else "FAIL"
        print(f"Input: {nums}, target={target}")
        print(f"  Brute:    floor={b[0]}, ceil={b[1]}")
        print(f"  Optimal:  floor={o[0]}, ceil={o[1]}")
        print(f"  Best:     floor={s[0]}, ceil={s[1]}")
        print(f"  Expected: floor={expected[0]}, ceil={expected[1]}  [{status}]\n")
